### PR TITLE
[Feature] Add SCA to Swift programming exercises

### PR DESCRIPTION
**Pull request by derLalla**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context

We want to give students a better feedback on their coding styles. So we are integrating SwiftLint into Artemis.

### Description

This PR adds SwiftLint as static code analysis for Swift programming exercises. For now, every rule is listed under the same category "Code Style". This is because SwiftLint doesn't integrate the category in the generated reports. This gets addressed in a later follow-up.
This PR also addresses the output of the build logs for Swift code. Duplicate logs are avoided.

### Steps for Testing

**This can be tested on TS3**

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a Swift Programming Exercise with SCA enabled
4. Click on Edit in Editor and select the Solution repo
5. Edit e.g. Client.swift and try to cause a SCA violation (For example add a semicolon at the end of a line)
6. Submit the code and wait for the build to finish
7. Open the Feedback section next to "Refresh files" and see that an entry was added with "CODE ISSUE"
8. If the build fails (e.g if you remove a bracket), verify a nice build log

To test on local server setup: 
You need to install the https://github.com/ls1intum/static-code-analysis-parser locally with `mvn install` and create a new bamboo reporter plugin https://github.com/ls1intum/bamboo-server-notification-plugin with `atlas-compile` and `atlas-run`. The resulting bamboo-server-1.3.4.jar can then be installed (updated) on your local bamboo instance.

### Test Coverage

BambooService.java: 82%
BambooBuildPlanService.java: 89%
StaticCodeAnalysisConfigurer.java: 100%
programming-exercise-update.component.ts: 71,36%

### Screenshots

image

image

image

SCA Feedback for the template exercise:
image

### CODE REVIEW

File: src/main/docker/bamboo/Dockerfile [hunk 1]
```diff
@@ -5,9 +5,10 @@ LABEL description="Bamboo pre-configured for Artemis"
 USER root
 ENV DEBIAN_FRONTEND noninteractive
 
-RUN apt-get update
-RUN apt-get install -y maven
-RUN apt-get install -y software-properties-common apt-utils
+RUN apt-get update && apt-get install -y \
+    maven \
+    software-properties-common \
+    apt-utils
```
Comment (derLalla) [inline src/main/docker/bamboo/Dockerfile:11]: which version are we using at the moment?

**Comment by stefanwaldhauser** [comment]
After hours of setting up I finally could test it and it works as expected :D Gj

**Review by stefanwaldhauser — APPROVED** [review]
Tested it locally and I could produce static code analysis for swift exercises . GJ

**Review by kloessst — APPROVED** [review]
Worked fine for me on TS3. Couldn't find any issues in the code. Well done!

File: src/main/docker/bamboo/Dockerfile [hunk 2]
```diff
@@ -5,9 +5,10 @@ LABEL description="Bamboo pre-configured for Artemis"
 USER root
 ENV DEBIAN_FRONTEND noninteractive
 
-RUN apt-get update
-RUN apt-get install -y maven
-RUN apt-get install -y software-properties-common apt-utils
+RUN apt-get update && apt-get install -y --no-install-recommends \
```
Comment (krusche) [inline src/main/docker/bamboo/Dockerfile:8]: why do we need to change this Dockerfile at all?
It also does not include dependencies to any other programming language

**Review by krusche — CHANGES_REQUESTED** [review]
The added tests are quite limited. I would expect more server and client test coverage in such a PR.
Also it does not make sense to me, to change the bamboo Dockerfile and add all Swift dependencies. The idea is that Bamboo uses docker build agents and does not need to install dependencies for all programming languages

**Comment by derLalla** [comment]
> The added tests are quite limited. I would expect more server and client test coverage in such a PR.
> Also it does not make sense to me, to change the bamboo Dockerfile and add all Swift dependencies. The idea is that Bamboo uses docker build agents and does not need to install dependencies for all programming languages

The Dockerfile needs to be changed in order for the local setup of bamboo to work. The build agents of bamboo are only used in the production environment. In the development/local environment (which is running within docker) those build agents are disabled. Therefore we need to install Swift and SwiftLint locally. 
Otherwise Swift exercises would not work on a local setup.

File: src/main/docker/bamboo/Dockerfile (hunk 2 above)
Comment (derLalla) [reply src/main/docker/bamboo/Dockerfile:8]: in case we wanted to use an updated version of maven, we need a way to tell docker to skip the cache and use the defined version. This is why it's a good practice to name the version of dependencies. This technique is known as “cache busting” (https://docs.docker.com/develop/develop-images/dockerfile_best-practices/)

**Comment by derLalla** [comment]
> The added tests are quite limited. I would expect more server and client test coverage in such a PR.
> Also it does not make sense to me, to change the bamboo Dockerfile and add all Swift dependencies. The idea is that Bamboo uses docker build agents and does not need to install dependencies for all programming languages

If it's not enough, I will write more tests. That's no problem at all.

**Review by tiffca — CHANGES_REQUESTED** [review]
I tried testing this on test server 3, but wasn't able to create the exercise. I deployed twice to confirm if the error was reproducible. 

Steps to reproduce: 
1. Create a swift programming exercise with SCA enabled and put 10% penalty 
2. Actual: Throws an ISE. Expected: it should create the exercise 

zone.js:3363 POST https://vmbruegge60.in.tum.de/api/programming-exercises/setup 500

**Review by krusche — CHANGES_REQUESTED** [review]
please add more tests and do not change the standard Dockerfile for bamboo.
You can instead create a second one that includes Swift, because I would like the standard one to stay as simple as possible

**Comment by derLalla** [comment]
> I tried testing this on test server 3, but wasn't able to create the exercise. I deployed twice to confirm if the error was reproducible.
> 
> Steps to reproduce:
> 
> 1. Create a swift programming exercise with SCA enabled and put 10% penalty
> 2. Actual: Throws an ISE. Expected: it should create the exercise
> 
> zone.js:3363 POST https://vmbruegge60.in.tum.de/api/programming-exercises/setup 500

this is strange. I have just did the same on ts3 and it worked for me: 

image

**Review by tiffca — DISMISSED** [review]
Tested successfully on ts3. It works if I create the exercise in a different course than my own.

**Review by tiffca — DISMISSED** [review]
tested successfully on ts3

**Review by airelawaleria — DISMISSED** [review]
Tested on ts3. Everything works. Faced the same issue as @tiffca . I cannot create exercises in most of the courses because of ISE. For me it only worked in the one particular course.

**Review by JohannesWeiss — APPROVED** [review]
Tested on ts3 as well and experienced the same problem with creation of exercises as mentioned before. Aside from that, the new feature worked as expected.

**Review by stefanwaldhauser — DISMISSED** [review]
Tested on TS3 with a new course and a new programming exercise and worked fine:
image

**Review by FrankeLukas — APPROVED** [review]
Works as advertised. One (probably unrelated) observation: Syntax errors aren't displayed really nicely, but that can be addressed in a different PR imo.
grafik

**Review by airelawaleria — APPROVED** [review]
Re-Tested on ts3. Works fine

**Review by JohannesWeiss — APPROVED** [review]
Retested on ts3, still works as intended. Re-Approve!

**Review by krusche — CHANGES_REQUESTED** [review]
I still miss some important server integration tests that simulate the calculation of results based on different configurations and different static code analysis reports.
The idea is that in the server integration test, we simulate new results (that would otherwise be sent by Bamboo or Jenkins) and check if those actually lead to correct scores and feedback elements.

Please add such tests and make sure they also verify the correctness of the whole approach for Java and Swift programming exercises!

**Comment by derLalla** [comment]
> Works as advertised. One (probably unrelated) observation: Syntax errors aren't displayed really nicely, but that can be addressed in a different PR imo.
> grafik

this is all done by SwiftLint, I don't have the possibility to change that

**Review by stefanwaldhauser — DISMISSED** [review]
Reapprove

**Review by tiffca — DISMISSED** [review]
Tested successfully on ts3

**Review by airelawaleria — APPROVED** [review]
Tested in the testing session with @marvinchimeka @simonzilker @JohannesWeiss @LudwigWeide @EneaGore   . Re-approve

**Review by simonzilker — DISMISSED** [review]
Works perfectly on ts3!

**Review by marvinchimeka — DISMISSED** [review]
Works as intended.

**Comment by derLalla** [comment]
> Tested on ts3 as well and experienced the same problem with creation of exercises as mentioned before. Aside from that, the new feature worked as expected.
> Edit: Tested still on version before last changes were added.

The problem with the ISE is not alone with Swift exercises, it occured with all other languages too.  
The error has something to do with a group not being existent:  
image

So it's not related to this PR ;)

**Review by JohannesWeiss — DISMISSED** [review]
Re-Approve

**Comment by FrankeLukas** [comment]
> this is all done by SwiftLint, I don't have the possibility to change that

I see, Swift developers are then likely already used to the layout too anyway.

**Review by FrankeLukas — DISMISSED** [review]
Re-approve.

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultJenkinsIntegrationTest.java [hunk 3]
```diff
@@ -38,20 +39,24 @@ void tearDown() {
...
     @WithMockUser(value = "student1", roles = "USER")
     public void shouldUpdateTestCasesAndResultScoreFromSolutionParticipationResult() {
+        programmingExerciseResultTestService.setup();
         var notification = ModelFactory.generateTestResultDTO(Constants.ASSIGNMENT_REPO_NAME, List.of("test1", "test2", "test4"), List.of());
         programmingExerciseResultTestService.shouldUpdateTestCasesAndResultScoreFromSolutionParticipationResult(notification);
     }
 
-    @Test
+    @ParameterizedTest
+    @EnumSource(value = ProgrammingLanguage.class, names = { "JAVA" })
```
Comment (krusche) [inline src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultJenkinsIntegrationTest.java:48]: please add swift here as well

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultJenkinsIntegrationTest.java [hunk 4]
```diff
@@ -38,20 +39,24 @@ void tearDown() {
...
     @WithMockUser(value = "student1", roles = "USER")
-    public void shouldStoreFeedbackForResultWithStaticCodeAnalysisReport() {
+    public void shouldStoreFeedbackForResultWithStaticCodeAnalysisReport(ProgrammingLanguage programmingLanguage) {
+        programmingExerciseResultTestService.setupForProgrammingLanguage(programmingLanguage);
         var notification = ModelFactory.generateTestResultDTO(Constants.ASSIGNMENT_REPO_NAME, List.of("test1"), List.of());
-        programmingExerciseResultTestService.shouldStoreFeedbackForResultWithStaticCodeAnalysisReport(notification);
+        programmingExerciseResultTestService.shouldStoreFeedbackForResultWithStaticCodeAnalysisReport(notification, ProgrammingLanguage.JAVA);
```
Comment (krusche) [inline src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultJenkinsIntegrationTest.java:53]:
```suggestion
        programmingExerciseResultTestService.shouldStoreFeedbackForResultWithStaticCodeAnalysisReport(notification, programmingLanguage);
```

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultBambooIntegrationTest.java [hunk 5]
```diff
@@ -38,27 +40,42 @@ void tearDown() {
...
+            assertThat(scaReports.size()).isEqualTo(4);
+            scaReports.get(0).getIssues().forEach(issue -> assertThat(issue.getCategory()).isNotNull());
+        }
+        programmingExerciseResultTestService.shouldStoreFeedbackForResultWithStaticCodeAnalysisReport(notification, programmingLanguage);
     }
 
     @Test
     @WithMockUser(value = "student1", roles = "USER")
     public void shouldStoreBuildLogsForSubmission() {
+        programmingExerciseResultTestService.setup();
```
Comment (krusche) [inline src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultBambooIntegrationTest.java:70]: why do you move the setup method to here?
The setup method should be called in the setup method above and not in each test case!

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultBambooIntegrationTest.java [hunk 6]
```diff
@@ -24,8 +28,6 @@
 
     @BeforeEach
     void setup() {
-        programmingExerciseResultTestService.setup();
```
Comment (krusche) [inline src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultBambooIntegrationTest.java:27]: please revert this change and keep the setup method here!

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseGitlabJenkinsIntegrationTest.java [hunk 7]
```diff
@@ -73,10 +73,13 @@ public void createProgrammingExercise_validExercise_bonusPointsIsNull() throws E
         programmingExerciseTestService.createProgrammingExercise_validExercise_bonusPointsIsNull();
     }
 
-    @Test
+    // TODO: Add template for Swift
+    @ParameterizedTest
+    @EnumSource(value = ProgrammingLanguage.class, names = { "JAVA" })
```
Comment (krusche) [inline src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseGitlabJenkinsIntegrationTest.java:78]: why is Swift still missing here? Please add it!

**Review by krusche — CHANGES_REQUESTED** [review]
@derLalla I still have some issues with the tests, see my inline comments.
In case the Swift files for Jenkins are missing intentionally, then please comment this accordingly that you will add this in the future.
With the new pipeline setup and the Docker agent support in Jenkins it should be really easy to add Swift exercises with static code analysis there as well. It would be ok to add this in a follow up PR.

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultBambooIntegrationTest.java (hunk 5 above)
Comment (derLalla) [reply src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultBambooIntegrationTest.java:70]: the setup method always creates a Java exercise, so we would need a parameterized @BeforeEach method, which always initializes the correct programming language exercise.   
Therefore I needed to split the setup() method into `setup()` and `setupForProgrammingLanguage(programmingLanguage)`. 

Otherwise, each time the testcase is non-Java, we would create an exercise with Java and an exercise for the specified prog. language.

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultBambooIntegrationTest.java (hunk 6 above)
Comment (derLalla) [reply src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultBambooIntegrationTest.java]: done in ba1458f587a5d5b3c9c465d0ab19e1c675d5d3f0

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultBambooIntegrationTest.java (hunk 5 above)
Comment (derLalla) [reply src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultBambooIntegrationTest.java]: resolved in ba1458f587a5d5b3c9c465d0ab19e1c675d5d3f0

**Comment by derLalla** [comment]
> @derLalla I still have some issues with the tests, see my inline comments.
> In case the Swift files for Jenkins are missing intentionally, then please comment this accordingly that you will add this in the future.
> With the new pipeline setup and the Docker agent support in Jenkins it should be really easy to add Swift exercises with static code analysis there as well. It would be ok to add this in a follow up PR.

The jenkins setup will be done in a follow-up :)

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultJenkinsIntegrationTest.java (hunk 4 above)
Comment (derLalla) [reply src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultJenkinsIntegrationTest.java]: resolved in bb131224aa3583ecda43327127046d7f590db25b e7d6a4a26d64212388c73f080885f6d0cf7345f3

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultJenkinsIntegrationTest.java (hunk 3 above)
Comment (derLalla) [reply src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseResultJenkinsIntegrationTest.java]: resolved in bb131224aa3583ecda43327127046d7f590db25b e7d6a4a26d64212388c73f080885f6d0cf7345f3

**Review by airelawaleria — APPROVED** [review]
Re-tested on ts3. Approve

**Review by JohannesWeiss — DISMISSED** [review]
Tested in testing session on ts3, works as intended!

**Review by simonzilker — DISMISSED** [review]
Works as described!

**Review by LudwigWeide — APPROVED** [review]
Tested in testing session. Approve.

**Review by marvinchimeka — APPROVED** [review]
Tested in testing session. Works fine

**Review by tiffca — APPROVED** [review]
Re-approve

**Review by airelawaleria — APPROVED** [review]
Re-approve

**Review by marvinchimeka — APPROVED** [review]
re-approve

**Review by simonzilker — APPROVED** [review]
Re-approve

**Review by krusche — APPROVED** [review]
